File: threadnote-mvp/Sources/Persistence/AppDatabase.swift
/// task003 — DatabaseMigrator v1 schema.
/// Creates all tables for the M1 persistence layer.

import Foundation
import GRDB

enum AppDatabase {

    static func makeDatabasePool(at url: URL) throws -> DatabasePool {
        let pool = try DatabasePool(path: url.path)
        try migrate(pool)
        return pool
    }

    // MARK: - Migrations

    static func migrate(_ writer: any DatabaseWriter) throws {
        var migrator = DatabaseMigrator()

        // Fail fast in development; remove for release builds.
        #if DEBUG
        migrator.eraseDatabaseOnSchemaChange = false
        #endif

        migrator.registerMigration("v1_initial_schema") { db in
            try db.execute(sql: """
                PRAGMA foreign_keys = ON;
                PRAGMA journal_mode = WAL;
            """)

            // threads
            try db.create(table: "threads", ifNotExists: true) { t in
                t.primaryKey("id", .text)
                t.column("title", .text).notNull()
                t.column("prompt", .text).notNull().defaults(to: "")
                t.column("goal_layer_json", .text).notNull().defaults(to: "{}")
                t.column("status", .text).notNull().defaults(to: "active")
                t.column("color", .text).notNull().defaults(to: "sky")
                t.column("created_at", .text).notNull()
                t.column("updated_at", .text).notNull()
                t.column("last_active_at", .text).notNull()
            }

            // entries
            try db.create(table: "entries", ifNotExists: true) { t in
                t.primaryKey("id", .text)
                t.column("thread_id", .text).references("threads", onDelete: .setNull)
                t.column("parent_entry_id", .text).references("entries", onDelete: .setNull)
                t.column("supersedes_entry_id", .text)
                t.column("kind", .text).notNull()
                t.column("body_json", .text).notNull().defaults(to: "{}")
                t.column("summary_text", .text).notNull().defaults(to: "")
                t.column("source_metadata_json", .text)
                t.column("author_type", .text).notNull().defaults(to: "user")
                t.column("inbox_state", .text).notNull().defaults(to: "unresolved")
                t.column("importance_score", .double)
                t.column("confidence_score", .double)
                t.column("session_id", .text)
                t.column("created_at", .text).notNull()
            }

            // object_mentions
            try db.create(table: "object_mentions", ifNotExists: true) { t in
                t.primaryKey("id", .text)
                t.column("entry_id", .text).notNull().references("entries", onDelete: .cascade)
                t.column("label", .text).notNull()
                t.column("kind", .text).notNull()
            }

            // entry_references
            try db.create(table: "entry_references", ifNotExists: true) { t in
                t.primaryKey("id", .text)
                t.column("entry_id", .text).notNull().references("entries", onDelete: .cascade)
                t.column("label", .text).notNull()
                t.column("target_kind", .text).notNull()
                t.column("target_id", .text)
            }

            // claims
            try db.create(table: "claims", ifNotExists: true) { t in
                t.primaryKey("id", .text)
                t.column("thread_id", .text).notNull().references("threads", onDelete: .cascade)
                t.column("origin_entry_id", .text).notNull()
                t.column("statement", .text).notNull()
                t.column("status", .text).notNull().defaults(to: "candidate")
                t.column("confidence_score", .double).notNull().defaults(to: 0.5)
                t.column("created_at", .text).notNull()
                t.column("updated_at", .text).notNull()
            }

            // anchors
            try db.create(table: "anchors", ifNotExists: true) { t in
                t.primaryKey("id", .text)
                t.column("thread_id", .text).notNull().references("threads", onDelete: .cascade)
                t.column("based_on_entry_id", .text)
                t.column("title", .text).notNull()
                t.column("core_question", .text).notNull()
                t.column("state_summary", .text).notNull()
                t.column("open_loops_json", .text).notNull().defaults(to: "[]")
                t.column("next_steps_json", .text).notNull().defaults(to: "[]")
                t.column("claim_ids_json", .text).notNull().defaults(to: "[]")
                t.column("evidence_entry_ids_json", .text).notNull().defaults(to: "[]")
                t.column("phase", .text).notNull()
                t.column("created_at", .text).notNull()
            }

            // discourse_relations
            try db.create(table: "discourse_relations", ifNotExists: true) { t in
                t.primaryKey("id", .text)
                t.column("source_entry_id", .text).notNull().references("entries", onDelete: .cascade)
                t.column("target_entry_id", .text).notNull().references("entries", onDelete: .cascade)
                t.column("kind", .text).notNull()
                t.column("confidence", .double).notNull().defaults(to: 1.0)
            }

            // tasks
            try db.create(table: "tasks", ifNotExists: true) { t in
                t.primaryKey("id", .text)
                t.column("thread_id", .text).notNull().references("threads", onDelete: .cascade)
                t.column("origin_entry_id", .text)
                t.column("title", .text).notNull()
                t.column("status", .text).notNull().defaults(to: "open")
                t.column("created_at", .text).notNull()
                t.column("updated_at", .text).notNull()
            }

            // app_metadata — migration version tracking
            try db.create(table: "app_metadata", ifNotExists: true) { t in
                t.primaryKey("key", .text)
                t.column("value", .text).notNull()
            }
        }

        migrator.registerMigration("v2_retrieval_layer") { db in
            // retrieval_documents — normalized recall units
            try db.create(table: "retrieval_documents", ifNotExists: true) { t in
                t.primaryKey("id", .text)
                t.column("owner_type", .text).notNull()   // entry/claim/anchor/resource
                t.column("owner_id", .text).notNull()
                t.column("thread_id", .text)
                t.column("title", .text).notNull().defaults(to: "")
                t.column("body", .text).notNull().defaults(to: "")
                t.column("metadata_json", .text).notNull().defaults(to: "{}")
                t.column("created_at", .text).notNull()
                t.column("updated_at", .text).notNull()
            }
            try db.create(indexOn: "retrieval_documents", columns: ["owner_id"])
            try db.create(indexOn: "retrieval_documents", columns: ["thread_id"])

            // retrieval_fts — FTS5 virtual table (content table mode)
            try db.execute(sql: """
                CREATE VIRTUAL TABLE IF NOT EXISTS retrieval_fts USING fts5(
                    title,
                    body,
                    content='retrieval_documents',
                    content_rowid='rowid',
                    tokenize='unicode61'
                )
            """)

            // Triggers to keep FTS index in sync with retrieval_documents
            try db.execute(sql: """
                CREATE TRIGGER IF NOT EXISTS retrieval_fts_ai
                AFTER INSERT ON retrieval_documents BEGIN
                    INSERT INTO retrieval_fts(rowid, title, body)
                    VALUES (new.rowid, new.title, new.body);
                END
            """)
            try db.execute(sql: """
                CREATE TRIGGER IF NOT EXISTS retrieval_fts_au
                AFTER UPDATE ON retrieval_documents BEGIN
                    INSERT INTO retrieval_fts(retrieval_fts, rowid, title, body)
                    VALUES ('delete', old.rowid, old.title, old.body);
                    INSERT INTO retrieval_fts(rowid, title, body)
                    VALUES (new.rowid, new.title, new.body);
                END
            """)
            try db.execute(sql: """
                CREATE TRIGGER IF NOT EXISTS retrieval_fts_ad
                AFTER DELETE ON retrieval_documents BEGIN
                    INSERT INTO retrieval_fts(retrieval_fts, rowid, title, body)
                    VALUES ('delete', old.rowid, old.title, old.body);
                END
            """)
        }

        try migrator.migrate(writer)
    }
}

File: threadnote-mvp/electron/src/infrastructure/persistence/schema/appDatabase.js
export const APP_DATABASE_MIGRATIONS = Object.freeze([
  {
    name: "v1_initial_schema",
    statements: [
      `CREATE TABLE IF NOT EXISTS threads (
        id TEXT PRIMARY KEY,
        title TEXT NOT NULL,
        prompt TEXT NOT NULL DEFAULT '',
        goal_layer_json TEXT NOT NULL DEFAULT '{}',
        status TEXT NOT NULL DEFAULT 'active',
        color TEXT NOT NULL DEFAULT 'sky',
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL,
        last_active_at TEXT NOT NULL
      )`,
      `CREATE TABLE IF NOT EXISTS entries (
        id TEXT PRIMARY KEY,
        thread_id TEXT,
        parent_entry_id TEXT,
        supersedes_entry_id TEXT,
        kind TEXT NOT NULL,
        body_json TEXT NOT NULL DEFAULT '{}',
        summary_text TEXT NOT NULL DEFAULT '',
        source_metadata_json TEXT,
        object_mentions_json TEXT NOT NULL DEFAULT '[]',
        references_json TEXT NOT NULL DEFAULT '[]',
        author_type TEXT NOT NULL DEFAULT 'user',
        inbox_state TEXT NOT NULL DEFAULT 'unresolved',
        importance_score REAL,
        confidence_score REAL,
        session_id TEXT,
        created_at TEXT NOT NULL
      )`,
      `CREATE TABLE IF NOT EXISTS claims (
        id TEXT PRIMARY KEY,
        thread_id TEXT NOT NULL,
        origin_entry_id TEXT NOT NULL,
        statement TEXT NOT NULL,
        status TEXT NOT NULL DEFAULT 'candidate',
        confidence_score REAL NOT NULL DEFAULT 0.5,
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL
      )`,
      `CREATE TABLE IF NOT EXISTS anchors (
        id TEXT PRIMARY KEY,
        thread_id TEXT NOT NULL,
        based_on_entry_id TEXT,
        title TEXT NOT NULL,
        core_question TEXT NOT NULL,
        state_summary TEXT NOT NULL,
        open_loops_json TEXT NOT NULL DEFAULT '[]',
        next_steps_json TEXT NOT NULL DEFAULT '[]',
        claim_ids_json TEXT NOT NULL DEFAULT '[]',
        evidence_entry_ids_json TEXT NOT NULL DEFAULT '[]',
        phase TEXT NOT NULL DEFAULT 'working',
        created_at TEXT NOT NULL
      )`,
      `CREATE TABLE IF NOT EXISTS discourse_relations (
        id TEXT PRIMARY KEY,
        source_entry_id TEXT NOT NULL,
        target_entry_id TEXT NOT NULL,
        kind TEXT NOT NULL,
        confidence REAL NOT NULL DEFAULT 1
      )`,
      `CREATE TABLE IF NOT EXISTS tasks (
        id TEXT PRIMARY KEY,
        thread_id TEXT NOT NULL,
        origin_entry_id TEXT,
        title TEXT NOT NULL,
        status TEXT NOT NULL DEFAULT 'open',
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL
      )`,
      `CREATE TABLE IF NOT EXISTS app_metadata (
        key TEXT PRIMARY KEY,
        value TEXT NOT NULL
      )`
    ]
  },
  {
    name: "v2_retrieval_layer",
    statements: [
      `CREATE TABLE IF NOT EXISTS retrieval_documents (
        id TEXT PRIMARY KEY,
        owner_type TEXT NOT NULL,
        owner_id TEXT NOT NULL,
        thread_id TEXT,
        title TEXT NOT NULL DEFAULT '',
        body TEXT NOT NULL DEFAULT '',
        metadata_json TEXT NOT NULL DEFAULT '{}',
        created_at TEXT NOT NULL,
        updated_at TEXT NOT NULL
      )`,
      "CREATE INDEX IF NOT EXISTS idx_retrieval_documents_owner_id ON retrieval_documents(owner_id)",
      "CREATE INDEX IF NOT EXISTS idx_retrieval_documents_thread_id ON retrieval_documents(thread_id)"
    ]
  },
  {
    name: "v3_memory_layer",
    statements: [
      `CREATE TABLE IF NOT EXISTS memory_records (
        id TEXT PRIMARY KEY,
        thread_id TEXT NOT NULL,
        scope TEXT NOT NULL,
        text TEXT NOT NULL,
        provenance TEXT NOT NULL,
        created_at TEXT NOT NULL
      )`,
      "CREATE INDEX IF NOT EXISTS idx_memory_records_thread_scope ON memory_records(thread_id, scope)"
    ]
  },
  {
    name: "v4_thread_ai_snapshots",
    statements: [
      `CREATE TABLE IF NOT EXISTS thread_ai_snapshots (
        thread_id TEXT PRIMARY KEY,
        content_fingerprint TEXT NOT NULL,
        headline TEXT NOT NULL DEFAULT '',
        blocks_json TEXT NOT NULL DEFAULT '[]',
        restart_note TEXT NOT NULL DEFAULT '',
        current_judgment TEXT NOT NULL DEFAULT '',
        open_loops_json TEXT NOT NULL DEFAULT '[]',
        next_action TEXT,
        recovery_lines_json TEXT NOT NULL DEFAULT '[]',
        synthesized_at REAL NOT NULL,
        model_id TEXT NOT NULL DEFAULT ''
      )`
    ]
  },
  {
    name: "v5_entry_status",
    statements: [
      "ALTER TABLE entries ADD COLUMN status TEXT NOT NULL DEFAULT 'open'",
      "ALTER TABLE entries ADD COLUMN status_metadata_json TEXT"
    ]
  },
  {
    name: "v6_entry_paging_indexes",
    statements: [
      "CREATE INDEX IF NOT EXISTS idx_entries_created_id ON entries(created_at DESC, id DESC)",
      "CREATE INDEX IF NOT EXISTS idx_entries_thread_created_id ON entries(thread_id, created_at DESC, id DESC)",
      "CREATE INDEX IF NOT EXISTS idx_entries_parent_created_id ON entries(parent_entry_id, created_at DESC, id DESC)"
    ]
  },
  {
    name: "v7_thread_materialized_versions",
    statements: [
      `CREATE TABLE IF NOT EXISTS thread_materialized_versions (
        thread_id TEXT PRIMARY KEY,
        memory_version INTEGER NOT NULL DEFAULT 0,
        retrieval_version INTEGER NOT NULL DEFAULT 0,
        updated_at TEXT NOT NULL
      )`,
      "CREATE INDEX IF NOT EXISTS idx_thread_materialized_versions_updated_at ON thread_materialized_versions(updated_at DESC)"
    ]
  },
  {
    name: "v8_thread_aggregates",
    statements: [
      `CREATE TABLE IF NOT EXISTS thread_aggregates (
        thread_id TEXT PRIMARY KEY,
        entry_count INTEGER NOT NULL DEFAULT 0,
        evidence_count INTEGER NOT NULL DEFAULT 0,
        source_count INTEGER NOT NULL DEFAULT 0,
        stable_claim_count INTEGER NOT NULL DEFAULT 0,
        stable_claim_updated_at TEXT,
        anchor_count INTEGER NOT NULL DEFAULT 0,
        latest_anchor_created_at TEXT,
        decided_count INTEGER NOT NULL DEFAULT 0,
        solved_count INTEGER NOT NULL DEFAULT 0,
        verified_count INTEGER NOT NULL DEFAULT 0,
        dropped_count INTEGER NOT NULL DEFAULT 0,
        updated_at TEXT NOT NULL
      )`,
      "CREATE INDEX IF NOT EXISTS idx_thread_aggregates_updated_at ON thread_aggregates(updated_at DESC)"
    ]
  }
]);

export function applyMigrations(database) {
  database.exec("PRAGMA journal_mode = WAL;");
  database.exec("PRAGMA foreign_keys = OFF;");
  database.exec("CREATE TABLE IF NOT EXISTS schema_migrations (name TEXT PRIMARY KEY, applied_at TEXT NOT NULL)");
  const existing = new Set(
    database.prepare("SELECT name FROM schema_migrations").all().map((row) => row.name)
  );

  for (const migration of APP_DATABASE_MIGRATIONS) {
    if (existing.has(migration.name)) {
      continue;
    }
    database.exec("BEGIN");
    try {
      for (const statement of migration.statements) {
        try {
          database.exec(statement);
        } catch (error) {
          const message = String(error?.message ?? "");
          if (/duplicate column name/i.test(message)) {
            continue;
          }
          throw error;
        }
      }
      database
        .prepare("INSERT INTO schema_migrations (name, applied_at) VALUES (?, ?)")
        .run(migration.name, new Date().toISOString());
      database.exec("COMMIT");
    } catch (error) {
      database.exec("ROLLBACK");
      throw error;
    }
  }
}
